bftengine/src/bftengine/PersistentStorage.hpp

```diff
@@ -0,0 +1,204 @@
+#include "ReplicaConfig.hpp"
+
+#include <vector>
+
+namespace bftEngine {
+namespace impl {
+class PrePrepareMsg;
+class CheckpointMsg;
+class ViewChangeMsg;
+class NewViewMsg;
+class FullCommitProofMsg;
+class PrepareFullMsg;
+class CommitFullMsg;
+
+// The PersistentStorage interface is used to write/read the concord-bft state 
+// to/from a persistent storage. In case the replica�s process is killed, 
+// crashed or restarted, this interface is used to restore the concord-bft 
+// state from the persistent storage. 
+// PersistentStorage is designed to only store data elements that are essential 
+// to reconstruct the internal in-memory data structures of concord-bft.
+// For simplicity and efficiency, some of the internal data elements are not 
+// written to the persistent storage (for example, messages with partial 
+// threshold signatures are not stored in persistent storage; if needed such 
+// messages are re-created and retransmitted by the replicas).
+
+class PersistentStorage {
+ public:
+  //////////////////////////////////////////////////////////////////////////
+  // Types
+  //////////////////////////////////////////////////////////////////////////
+
+  struct DescriptorOfLastExitFromView {
+    // view >= 0
+    ViewNum view;
+
+    // lastStable >= 0
+    SeqNum lastStable;
+
+    // lastExecuted >= lastStable
+    SeqNum lastExecuted;
+
+    // elements.size() <= kWorkWindowSize
+    // The messages in elements[i] may be null
+    std::vector<ViewsManager::PrevViewInfo> elements;
+  };
+
+  struct DescriptorOfLastNewView {
+    // view >= 1
+    ViewNum view;
+
+    // newViewMsg != nullptr
+    NewViewMsg* newViewMsg;
+
+    // viewChangeMsgs.size() == 2*F + 2*C + 1
+    // The messages in viewChangeMsgs will never be null
+    std::vector<ViewChangeMsg*> viewChangeMsgs;
+
+    // maxSeqNumTransferredFromPrevViews >= 0
+    SeqNum maxSeqNumTransferredFromPrevViews;
+  };
+
+  struct DescriptorOfLastExecution {
+    // executedSeqNum >= 1
+    SeqNum executedSeqNum;
+
+    // 1 <= validRequests.numOfBits() <= maxNumOfRequestsInBatch
+    Bitmap validRequests;
+  };
+
+  //////////////////////////////////////////////////////////////////////////
+  // Transactions management
+  //////////////////////////////////////////////////////////////////////////
+
+  // begin reentrant write-only transaction
+  // returns the number of nested transactions
+  virtual uint8_t beginWriteTran() = 0;
+
+  // end reentrant write-only transaction
+  // returns the number of remaining nested transactions
+  virtual uint8_t endWriteTran() = 0;
+
+  // return true IFF write-only transactions are running now
+  virtual bool isInWriteTran() const = 0;
+
+  //////////////////////////////////////////////////////////////////////////
+  // Update methods (should only be used in write-only transactions)
+  //////////////////////////////////////////////////////////////////////////
+
+  virtual void setReplicaConfig(ReplicaConfig config) = 0;
```

Comment:
Is there any reason to pass this structure by value?

Reply:
For ReplicaConfig , it doesn't really matter.
However, I agree that it will be better to update:
 virtual void setReplicaConfig(const ReplicaConfig& config) = 0;